drivers/led/issi/is31fl3729.c

```diff
@@ -0,0 +1,211 @@
+// Storing them like this is optimal for I2C transfers to the registers.
+uint8_t g_pwm_buffer[IS31FL3729_DRIVER_COUNT][IS31FL3729_PWM_REGISTER_COUNT];
+bool    g_pwm_buffer_update_required[IS31FL3729_DRIVER_COUNT] = {false};
+
+uint8_t g_scaling_registers[IS31FL3729_DRIVER_COUNT][IS31FL3729_SCALING_REGISTER_COUNT];
+bool    g_scaling_registers_update_required[IS31FL3729_DRIVER_COUNT] = {false};
+
+void is31fl3729_write_register(uint8_t addr, uint8_t reg, uint8_t data) {
+#if IS31FL3729_I2C_PERSISTENCE > 0
+    for (uint8_t i = 0; i < IS31FL3729_I2C_PERSISTENCE; i++) {
+        if (i2c_write_register(addr << 1, reg, &data, 1, IS31FL3729_I2C_TIMEOUT) == I2C_STATUS_SUCCESS) break;
+    }
+#else
+    i2c_write_register(addr << 1, reg, &data, 1, IS31FL3729_I2C_TIMEOUT);
+#endif
+}
+
+void is31fl3729_write_pwm_buffer(uint8_t addr, uint8_t index) {
+    // Transmit PWM registers in 9 transfers of 16 bytes.
+
+    // Iterate over the pwm_buffer contents at 16 byte intervals.
+    for (uint8_t i = 0; i <= IS31FL3729_PWM_REGISTER_COUNT; i += 16) {
+#if IS31FL3729_I2C_PERSISTENCE > 0
+        for (uint8_t j = 0; j < IS31FL3729_I2C_PERSISTENCE; j++) {
+            if (i2c_write_register(addr << 1, i, g_pwm_buffer[index] + i, 16, IS31FL3729_I2C_TIMEOUT) == I2C_STATUS_SUCCESS) break;
+        }
+#else
+        i2c_write_register(addr << 1, i, g_pwm_buffer[index] + i, 16, IS31FL3729_I2C_TIMEOUT);
+#endif
+    }
+}
+
+void is31fl3729_init_drivers(void) {
+    i2c_init();
+
+    is31fl3729_init(IS31FL3729_I2C_ADDRESS_1);
+#if defined(IS31FL3729_I2C_ADDRESS_2)
+    is31fl3729_init(IS31FL3729_I2C_ADDRESS_2);
+#    if defined(IS31FL3729_I2C_ADDRESS_3)
+    is31fl3729_init(IS31FL3729_I2C_ADDRESS_3);
+#        if defined(IS31FL3729_I2C_ADDRESS_4)
+    is31fl3729_init(IS31FL3729_I2C_ADDRESS_4);
+#        endif
+#    endif
+#endif
+
+    for (int i = 0; i < IS31FL3729_LED_COUNT; i++) {
+        is31fl3729_set_scaling_register(i, 0xFF, 0xFF, 0xFF);
+    }
+
+    is31fl3729_update_scaling_registers(IS31FL3729_I2C_ADDRESS_1, 0);
+#if defined(IS31FL3729_I2C_ADDRESS_2)
+    is31fl3729_update_scaling_registers(IS31FL3729_I2C_ADDRESS_2, 1);
+#    if defined(IS31FL3729_I2C_ADDRESS_3)
+    is31fl3729_update_scaling_registers(IS31FL3729_I2C_ADDRESS_3, 2);
+#        if defined(IS31FL3729_I2C_ADDRESS_4)
+    is31fl3729_update_scaling_registers(IS31FL3729_I2C_ADDRESS_4, 3);
+#        endif
+#    endif
+#endif
+}
+
+void is31fl3729_init(uint8_t addr) {
+    // In order to avoid the LEDs being driven with garbage data
+    // in the LED driver's PWM registers, shutdown is enabled last.
+    // Set up the mode and other settings, clear the PWM registers,
+    // then disable software shutdown.
+
+    // Set Pull up & Down for SWx CSy
+    is31fl3729_write_register(addr, IS31FL3729_REG_PULLDOWNUP, IS31FL3729_PULLDOWNUP);
+
+    // Set Spread Spectrum Register if applicable
+    is31fl3729_write_register(addr, IS31FL3729_REG_SPREAD_SPECTRUM, IS31FL3729_SPREAD_SPECTRUM);
```

Comment:
```suggestion
    is31fl3729_write_register(addr, IS31FL3729_REG_SPREAD_SPECTRUM, ((IS31FL3729_SPREAD_SPECTRUM & 0b1) << 4) | ((IS31FL3729_SPREAD_SPECTRUM_RANGE & 0b11) << 2) | (IS31FL3729_SPREAD_SPECTRUM_CYCLE_TIME & 0b11));
```

Reply:
Thanks, I've just push a new commit to resolved this.